torch/csrc/Device.cpp

```diff
@@ -0,0 +1,217 @@
+{
+  auto type = (PyTypeObject*)&THPDeviceType;
+  auto self = THPObjectPtr{type->tp_alloc(type, 0)};
+  if (!self) throw python_error();
+  auto self_ = reinterpret_cast<THPDevice*>(self.get());
+  self_->device_type = device_type;
+  self_->device_index = device_index;
+  self_->is_default = is_default;
+  return self.release();
+}
+
+static const char* cuda_str = "cuda";
+static const char* cpu_str = "cpu";
+
+static inline const char* deviceTypeString(torch::DeviceType device_type) {
+  switch (device_type) {
+    case torch::DeviceType::CUDA:
+      return cuda_str;
+    case torch::DeviceType::CPU:
+      return cpu_str;
+    default:
+      throw std::runtime_error("unexpected device type");
+  }
+}
+
+PyObject *THPDevice_repr(THPDevice *self)
+{
+  std::ostringstream oss;
+  oss << "Device(device_type=\'" << deviceTypeString(self->device_type) << "\'";
+  if (!self->is_default) {
+    oss << ", device_index=" << self->device_index;
+  }
+  oss << ")";
+  return THPUtils_packString(oss.str().c_str());
+}
+
+PyObject *THPDevice_str(THPDevice*self)
+{
+  std::ostringstream oss;
+  if (!self->is_default) {
+    oss << deviceTypeString(self->device_type) << ":" << self->device_index;
+  } else {
+    oss << deviceTypeString(self->device_type);
+  }
+  return THPUtils_packString(oss.str().c_str());
+}
+
+PyObject *THPDevice_pynew(PyTypeObject *type, PyObject *args, PyObject *kwargs)
+{
+  HANDLE_TH_ERRORS
+  static torch::PythonArgParser parser({
+    "Device(Device device)",
+    "Device(String device_type, int64_t? device_index=-1)"
+  });
+  torch::ParsedArgs<2> parsed_args;
+  auto r = parser.parse(args, kwargs, parsed_args);
+  if (r.idx == 0) {
+    auto device = r.device(0);
+    return THPDevice_New(device.device_type, device.device_index, device.is_default);
+  } else if (r.idx == 1) {
+    auto as_device = r.device(0);  // this works, because device can take strings
+    auto device_type = r.string(0);
+    if (!as_device.is_default) {
+      throw std::runtime_error("device_type must not include index because index is passed explicitly " + device_type);
+    }
+
+    auto is_default = r.isNone(1);
+    auto device_index = r.toInt64WithDefault(1, -1);
+    // make sure this is constructible
+    auto device = torch::Device(as_device.device_type, device_index, is_default);
+    return THPDevice_New(device.device_type, device.device_index, device.is_default);
+  }
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject *THPDevice_type(THPDevice *self)
+{
+  HANDLE_TH_ERRORS
+  return THPUtils_packString(deviceTypeString(self->device_type));
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject *THPDevice_index(THPDevice *self)
+{
+  HANDLE_TH_ERRORS
+  if (self->is_default) {
+    Py_RETURN_NONE;
+  } else {
+    return THPUtils_packInt64(self->device_index);
+  }
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject *THPDevice_cuda_index(THPDevice *self)
```

Comment:
Do we really want this to be part of the public API? 😕 Will we have `X_index` for every device kind?

Reply:
I put this in there after talking to @teng-li; DistributedDataParallel takes lists of integers, this makes it easy to have the same exact behavior previously but also allowing strings/devices.  I assume there are more usages for that sort of thing.

So, no, wasn't planning to make `X_index` for every device kind, just to have this to ease the transition.  Maybe I should make this a standalone function in `backcompat` though?

Reply:
Not sure what `backcompat` is, but I wouldn't make it official API, if it's going to be different for other devices

Reply:
`torch.utils.backcompat` -- it currently has stuff about settings warnings for broadcasting/keepdim, but this is backcompat related.